api/src/main/java/org/apache/iceberg/transforms/PartitionSpecVisitor.java

```diff
@@ -26,20 +26,94 @@
 import org.apache.iceberg.relocated.com.google.common.collect.Lists;
 
 public interface PartitionSpecVisitor<T> {
-  T identity(String sourceName, int sourceId);
+  default T identity(int fieldId, String sourceName, int sourceId) {
+    return identity(sourceName, sourceId);
+  }
+
+  default T identity(String sourceName, int sourceId) {
+    throw new UnsupportedOperationException("Identity transform is not supported");
+  }
+
+  default T bucket(int fieldId, String sourceName, int sourceId, int width) {
+    return bucket(sourceName, sourceId, width);
+  }
+
+  default T bucket(String sourceName, int sourceId, int width) {
+    throw new UnsupportedOperationException("Bucket transform is not supported");
+  }
 
-  T bucket(String sourceName, int sourceId, int width);
+  default T truncate(int fieldId, String sourceName, int sourceId, int width) {
+    return truncate(sourceName, sourceId, width);
+  }
 
-  T truncate(String sourceName, int sourceId, int width);
+  default T truncate(String sourceName, int sourceId, int width) {
+    throw new UnsupportedOperationException("Truncate transform is not supported");
+  }
 
-  T year(String sourceName, int sourceId);
+  default T year(int fieldId, String sourceName, int sourceId) {
+    return year(sourceName, sourceId);
+  }
 
-  T month(String sourceName, int sourceId);
+  default T year(String sourceName, int sourceId) {
+    throw new UnsupportedOperationException("Year transform is not supported");
+  }
 
-  T day(String sourceName, int sourceId);
+  default T month(int fieldId, String sourceName, int sourceId) {
+    return month(sourceName, sourceId);
+  }
+
+  default T month(String sourceName, int sourceId) {
+    throw new UnsupportedOperationException("Month transform is not supported");
+  }
 
-  T hour(String sourceName, int sourceId);
+  default T day(int fieldId, String sourceName, int sourceId) {
+    return day(sourceName, sourceId);
+  }
+
+  default T day(String sourceName, int sourceId) {
+    throw new UnsupportedOperationException("Day transform is not supported");
+  }
+
+  default T hour(int fieldId, String sourceName, int sourceId) {
+    return hour(sourceName, sourceId);
+  }
+
+  default T hour(String sourceName, int sourceId) {
+    throw new UnsupportedOperationException("Hour transform is not supported");
+  }
+
+  default T alwaysNull(int fieldId, String sourceName, int sourceId) {
```

Comment:
nit: should we add `default T alwaysNull(String sourceName, int sourceId)` as well?

Reply:
No, the methods without fieldId are only there for backward compatibility.